src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs

```diff
@@ -516,7 +517,37 @@ public bool IsCompleted
             /// <param name="continuation">The delegate to invoke.</param>
             public void OnCompleted(Action continuation)
             {
-                Task.Factory.StartNew(continuation, CancellationToken.None, TaskCreationOptions.None, this.scheduler);
+                if (this.scheduler == TaskScheduler.Default)
+                {
+                    ThreadPool.QueueUserWorkItem(state => ((Action)state)(), continuation);
+                }
+                else
+                {
+                    Task.Factory.StartNew(continuation, CancellationToken.None, TaskCreationOptions.None, this.scheduler);
+                }
+            }
+
+            /// <summary>
+            /// Schedules a continuation to execute using the specified task scheduler
+            /// without capturing the ExecutionContext.
+            /// </summary>
+            /// <param name="continuation">The action.</param>
+            [SecurityCritical]
+            public void UnsafeOnCompleted(Action continuation)
+            {
+                if (this.scheduler == TaskScheduler.Default)
+                {
+#if THREADPOOL
+                    ThreadPool.UnsafeQueueUserWorkItem(state => ((Action)state)(), continuation);
+#else
+                    ThreadPool.QueueUserWorkItem(state => ((Action)state)(), continuation);
+#endif
+                }
+                else
+                {
+                    // There is no API for scheduling a Task without capturing the ExecutionContext.
```

Comment:
You could wrap it with ExecutionContext.SuppressFlow/RestoreFlow if you wanted, e.g.
https://github.com/dotnet/corefx/blob/ec5fb4162c8376cf8ced69f1d2442c586e9197f3/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs#L289-L303
but you'd want to measure to see whether it was actually a win; there are some tradeoffs in costs.  My guess is that on desktop it would not be a beneficial change.

Reply:
Thanks for the tip. The change you're suggesting I consider would have an allocation win for sure, I think, is that right? It's just the perf cost of suppressing/restoring flow that might negate any perf win of not capturing/applying it, I think?

Also, the second PR will include some more complex awaiters, including one with `SynchronizationContext.Post`  calls and several code branches in multiple classes, and I wasn't sure I wanted to (or even could) suppress flow in all cases. `SuppressFlow` looks like an interesting option both to localize the switching off of the flow, and to enable skipping it on those methods that don't offer a way to skip it. Does this work generally?